src/components/validations/with-validation.hoc.js

```diff
@@ -12,122 +13,137 @@ const withValidation = (WrappedComponent) => {
+        ...acc,
+        [type]: PropTypes.oneOfType([ // The info validations that should be run against the value
+          PropTypes.func,
+          PropTypes.arrayOf(
+            PropTypes.shape({
+              message: PropTypes.func,
+              validate: PropTypes.func
+            })
+          ),
+          PropTypes.arrayOf(PropTypes.func)
+        ])
+      }), {})
     };
 
-    static defaultProps = {
-      validations: [],
-      warnings: [],
-      info: []
-    }
+    static defaultProps = Object.values(VALIDATION_TYPES).reduce((acc, type) => ({
+      ...acc,
+      [type]: []
+    }), {});
 
-    state = {
-      hasError: false,
-      hasWarning: false,
-      hasInfo: false
-    };
+    state = Object.keys(VALIDATION_TYPES).reduce((acc, type) => ({
+      ...acc,
+      [`${type}Message`]: undefined
+    }), {});
 
     componentDidMount() {
-      if (this.checkValidations(['info', 'warnings', 'validations'])) {
+      if (this.checkValidations()) {
         this.context.addInput(this.props.name, this.validate);
       }
     }
 
     componentWillUnmount() {
-      if (this.context.removeInput) this.context.removeInput(this.props.name);
+      if (this.context && this.context.removeInput) this.context.removeInput(this.props.name);
     }
 
     componentDidUpdate(prevProps) {
-      if (this.isUpdatedValidationProps(prevProps) && this.checkValidations(['info', 'warnings', 'validations'])) {
+      if (this.isUpdatedValidationProps(prevProps) && this.checkValidations()) {
         this.context.addInput(this.props.name, this.validate);
       }
+
+      if (prevProps.value !== this.props.value) {
+        this.validate();
+      }
     }
 
     isUpdatedValidationProps(prevProps) {
-      const { validations, warnings, info } = this.props;
-      return validations !== prevProps.validations || warnings !== prevProps.warnings || info !== prevProps.info;
+      let updated = false;
+      const validationTypes = Object.values(VALIDATION_TYPES);
+      validationTypes.forEach((type) => {
+        if (this.props[type] !== prevProps[type]) updated = true;
+      });
+      return updated;
     }
 
-    checkValidations(types) {
-      if (!this.context.addInput) return false;
+    checkValidations() {
+      if (!this.context || !this.context.addInput) return false;
 
-      let hasValidations = true;
-      types.forEach((type) => {
-        if ((Array.isArray(this.props[type]) && this.props[type].length === 0)
-          || typeof this.props[type] === 'undefined') {
-          hasValidations = false;
-        } else {
+      const types = Object.keys(VALIDATION_TYPES);
+      let hasValidations = false;
+      types.forEach((validationType) => {
+        const type = VALIDATION_TYPES[validationType];
+        const isArray = Array.isArray(this.props[type]);
+        if ((isArray && this.props[type].length)
+          || (!isArray && typeof this.props[type] !== 'undefined')) {
```

Comment:
`this.props[type]` could be a const

Reply:
these conditions could also be extracted to consts to increase readability, e.g.: `if (isPopulatedArray || isValidationPropDefined)`